src/multiversioning.zig

```diff
@@ -911,16 +930,34 @@ pub const Multiversion = struct {
         result: IO.ReadError!usize,
     ) void {
         assert(self.stage == .source_read);
-
-        posix.close(self.source_fd.?);
-        self.source_fd = null;
+        assert(self.source_fd != null);
+        assert(self.source_offset != null);
+
+        defer {
+            if (self.stage != .source_read) {
+                assert(self.stage == .err or self.stage == .ready);
+                assert(self.source_fd != null);
+                assert(self.source_offset != null);
+
+                posix.close(self.source_fd.?);
+                self.source_offset = null;
+                self.source_fd = null;
+            }
+        }
 
         const bytes_read = result catch |e| return self.handle_error(e);
-        const source_buffer = self.source_buffer[0..bytes_read];
+        self.source_offset.? += bytes_read;
+        assert(self.source_offset.? <= self.source_buffer.len);
```

Comment:
Hm, so what happens here if we read a binary that's larger than our buffer?

It seems that:

* we'll read right to the end of the buffer
* and then we'll do another `read` syscall passing-in an empty slice
* and that'll return 0 bytes read
* so we'll interpret that as a file of just the right size

Should we perhaps explicitly treat that as an error?

But I think I also have an argument for why this is OK: we don't have a guarantee that we read the entire binary anyway. Like, if someone `cp`s binary by doing writes, rather than an atomic rename, we totally can get truncation. So we have to deal with truncated binaries anyway, and we do, by verifying checksum:

https://github.com/tigerbeetle/tigerbeetle/blob/cd8f4e556c0cdb9726e838f21a73bc5a078fcb71/src/multiversioning.zig#L966-L971

So I'd say the logic here is correct, but is correct in an un-obvious way, and it's worth adding something like

```
// This could be a truncated file, but it'll get rejected 
// when we verify the checksum.
maybe(self.source_offset.? == self.source_buffer.len);
```